lib/scholar/cluster/affinity_propagation.ex

```diff
@@ -0,0 +1,321 @@
+          [-6062.0, -939.5, -966.0, -913.0],
+          [-5310.0, -966.0, -939.5, -3.0],
+          [-5459.0, -913.0, -3.0, -939.5]
+        ]
+      ),
+      Nx.tensor(
+        [
+          [0.0, -6062.0, -5310.0, -5459.0],
+          [-6062.0, 0.0, -966.0, -913.0],
+          [-5310.0, -966.0, 0.0, -3.0],
+          [-5459.0, -913.0, -3.0, 0.0]
+        ]
+      ),
+      Nx.tensor(
+        [1, 0, 0, 1], type: :u8
+      ),
+      Nx.tensor(
+        [
+          [12, 5, 78, 2],
+          [1, 5, 7, 32],
+          [1, 3, 6, 1],
+          [1, 2, 5, 2]
+        ]
+      )}
+  """
+  deftransform fit(data, opts \\ []) do
+    opts = NimbleOptions.validate!(opts, @opts_schema)
+    opts = Keyword.update(opts, :self_preference, false, fn x -> x end)
+    seed = Keyword.get_lazy(opts, :seed, &System.system_time/0)
+    fit_n(data, seed, NimbleOptions.validate!(opts, @opts_schema))
+  end
+
+  defnp fit_n(data, seed, opts) do
+    iterations = opts[:iterations]
+    damping_factor = opts[:damping_factor]
+    self_preference = opts[:self_preference]
+
+    {initial_a, initial_r, s, affinity_matrix} =
+      initialize_matrices(data, self_preference: self_preference)
+
+    {n, _} = Nx.shape(initial_a)
+    key = Nx.Random.key(seed)
+    {normal, _new_key} = Nx.Random.normal(key, 0, 1, shape: {n, n}, type: Nx.type(s))
+
+    s =
+      s +
+        normal *
+          (Nx.Constants.smallest_positive_normal(Nx.type(s)) * 100 + 2.220446049250313e-16 * s)
+
+    range = Nx.iota({n})
+
+    {_final_a, _final_r, diagonals, _, _} =
+      while {a = initial_a, r = initial_r,
+             _diagonals = Nx.broadcast(Nx.tensor(0, type: {:u, 8}), {n}), s = s, i = 0},
+            i < iterations do
+        temp = a + s
+
+        indices = Nx.argmax(temp, axis: 1)
+
+        y = Nx.reduce_max(temp, axes: [1])
+
+        neg_inf = Nx.Constants.neg_infinity(Nx.Type.to_floating(Nx.type(a)))
+        inf = Nx.Constants.infinity(Nx.Type.to_floating(Nx.type(a)))
+
+        neg_infinities = Nx.broadcast(neg_inf, {n})
+
+        max_indices =
+          Nx.stack([range, indices])
+          |> Nx.transpose()
+
+        temp = Nx.indexed_put(temp, max_indices, neg_infinities)
+
+        y2 = Nx.reduce_max(temp, axes: [1])
+
+        temp = s - Nx.new_axis(y, -1)
+
+        temp =
+          Nx.indexed_put(
+            temp,
+            max_indices,
+            Nx.flatten(Nx.gather(s, max_indices) - y2)
+          )
+
+        temp = temp * (1 - damping_factor)
+
+        r = r * damping_factor + temp
+
+        temp = Nx.max(r, 0)
+
+        temp = Nx.put_diagonal(temp, Nx.take_diagonal(r))
+
+        temp = temp - Nx.sum(temp, axes: [0])
+
+        a_change = Nx.take_diagonal(temp)
+
+        temp = Nx.clip(temp, 0, inf)
+
+        temp = Nx.put_diagonal(temp, a_change)
+
+        temp = temp * (1 - damping_factor)
+
+        a = a * damping_factor - temp
+
+        diagonals = Nx.take_diagonal(a) + Nx.take_diagonal(r) > 0
```

Comment:
Since diagonal does not seem to be used throughout the loop, I think you can compute this outside of the `while` by using the last `a` and `r`?